cpp/src/arrow/dataset/test_util.h

```diff
@@ -631,12 +646,92 @@ class FileFormatScanMixin : public FileFormatFixtureMixin<FormatHelper>,
     for (auto maybe_batch : PhysicalBatches(fragment)) {
       ASSERT_OK_AND_ASSIGN(auto batch, maybe_batch);
       row_count += batch->num_rows();
-      AssertSchemaEqual(*batch->schema(), *expected_schema,
-                        /*check_metadata=*/false);
+      ASSERT_THAT(
+          batch->schema()->fields(),
+          ::testing::UnorderedPointwise(PointeesEquals(), expected_schema->fields()))
+          << "EXPECTED:\n"
+          << expected_schema->ToString() << "\nACTUAL:\n"
+          << batch->schema()->ToString();
     }
 
     ASSERT_EQ(row_count, expected_rows());
   }
+  void TestScanProjectedNested(bool fine_grained_selection = false) {
+    auto f32 = field("f32", float32());
+    auto f64 = field("f64", float64());
+    auto i32 = field("i32", int32());
+    auto i64 = field("i64", int64());
+    auto struct1 = field("struct1", struct_({f32, i32}));
+    auto struct2 = field("struct2", struct_({f64, i64, struct1}));
+    this->SetSchema({struct1, struct2, f32, f64, i32, i64});
+    this->ProjectNested({".struct1.f32", ".struct2.struct1", ".struct2.struct1.f32"});
+    this->SetFilter(equal(field_ref(FieldRef("struct2", "i64")), literal(0)));
+
+    std::shared_ptr<Schema> expected_schema;
+    if (fine_grained_selection) {
+      // Some formats, like Parquet, let you pluck only a part of a complex type
+      expected_schema = schema({
+          field("struct1", struct_({f32})),
+          field("struct2", struct_({i64, struct1})),
+      });
+    } else {
+      expected_schema = schema({struct1, struct2});
+    }
```

Comment:
If I'm understanding this I'm not sure I like it.  I would expect the resulting schema to be the same regardless of whether the underlying format supported partial projection or not.  For formats that don't support partial projection I would expect it would be simulated by a full read and then a cast.

Reply:
The overall schema will be the same once we pass through projection, i.e. the cast is done in the scanner instead of inside every file format. However, the tests here are reading from the fragment directly to check the physical schema, instead of the post-projection schema. I'll make sure both cases are covered in tests, though.

Reply:
Ah…this is a little problematic since filtering/projection cast to the dataset schema first, and then we run into ARROW-1888 again as a result. I might go implement that first since this PR becomes a lot more useful with that.

Reply:
(And, well, ARROW-1888 is a little easier with ARROW-7051…)

Reply:
Yes, projection is not inside the exec plan.  Also, projection doesn't occur until near the end of the exec plan (e.g. the filter step runs on the unprojected data).  So it is important for the scan to do its own internal projection to the dataset schema.

If the blocker is ARROW-1888 I think it would be fine to implement this as-is with comments for a follow-up JIRA next to the test behavior we expect to change.

Reply:
No tests will change (other than the one marked already), but as-is, you can't scan a Parquet dataset and project a nested field (since when we project from the specific schema to the dataset schema, we'll fail).

Reply:
Alright, I've disabled fine-grained projection for now and marked it with ARROW-1888.